core/config/config_helper.cpp

```diff
@@ -62,8 +63,87 @@ parse_or_get_factory<const stop::CriterionFactory>(const pnode& config,
+                       const type_descriptor& td) {
+        pnode time_config{{{"time_limit", config.get("time")}}};
+        return configure_time(time_config, context, td);
+    };
+    auto map_iteration = [](const pnode& config, const registry& context,
+                            const type_descriptor& td) {
+        pnode iter_config{{{"max_iters", config.get("iteration")}}};
+        return configure_iter(iter_config, context, td);
+    };
+    auto create_residual_mapping = [](const std::string& key,
+                                      const std::string& baseline,
+                                      auto configure_fn) {
+        return std::make_pair(
+            key, [=](const pnode& config, const registry& context,
+                     const type_descriptor& td) {
+                pnode res_config{{{"baseline", pnode{baseline}},
+                                  {"reduction_factor", config.get(key)}}};
+                return configure_fn(res_config, context, td);
+            });
+    };
+    std::map<
+        std::string,
+        std::function<deferred_factory_parameter<gko::stop::CriterionFactory>(
+            const pnode&, const registry&, type_descriptor)>>
+        criterion_map{
+            {{"time", map_time},
+             {"iteration", map_iteration},
+             create_residual_mapping("relative_residual_norm", "rhs_norm",
+                                     configure_residual),
+             create_residual_mapping("initial_residual_norm", "initial_resnorm",
+                                     configure_residual),
+             create_residual_mapping("absolute_residual_norm", "absolute",
+                                     configure_residual),
+             create_residual_mapping("relative_implicit_residual_norm",
+                                     "rhs_norm", configure_implicit_residual),
+             create_residual_mapping("initial_implicit_residual_norm",
+                                     "initial_resnorm",
+                                     configure_implicit_residual),
+             create_residual_mapping("absolute_implicit_residual_norm",
+                                     "absolute", configure_implicit_residual)}};
+
+    std::vector<deferred_factory_parameter<const stop::CriterionFactory>> res;
+    for (const auto& it : config.get_map()) {
+        res.emplace_back(criterion_map.at(it.first)(config, context, td));
+    }
+    return res;
+}
+
+
+std::vector<deferred_factory_parameter<const stop::CriterionFactory>>
+parse_or_get_criteria(const pnode& config, const registry& context,
+                      const type_descriptor& td)
+{
+    if (config.get_tag() == pnode::tag_t::array) {
+        return parse_or_get_factory_vector<const stop::CriterionFactory>(
+            config, context, td);
+    }
+
+    if (config.get_tag() == pnode::tag_t::map) {
+        return parse_minimal_criteria(config, context, td);
```

Comment:
```suggestion
        auto updated = config::update_type(td);
        return parse_minimal_criteria(config, context, updated);
```
It is to support no valuetype available outside

Reply:
I'm not sure what you mean here. I can specify the value type of the residual nom criterion, as you can see in the tests.

Reply:
It's based on the type_descriptor from the outer loop.
I mean something like
```
"stop": {
  "value_type": "float64",
  "residual_norm": ...
}
```
in case no precision information from outside or want to specify certain precision for stop.

Reply:
I will try a bit to make it work. If it's a reasonable effort I will adapt it.

Reply:
It was pretty simple to add, so it's now available.